epuck_ros2_camera/src/camera.cpp

```diff
@@ -87,63 +116,122 @@ class CameraPublisher : public rclcpp::Node {
-      if (!mV4l2Initialized) {
-        pipuck_v4l2_init();
-        mV4l2Initialized = true;
-      }
-      pipuck_v4l2_capture(&mCapturedImage);
-    } else if (mV4l2Initialized) {
-      pipuck_v4l2_deinit();
-      mV4l2Initialized = false;
+      // Capture an image
+      initV4l2();
+      pipuck_v4l2_capture(&(mPipuckMmalRgb.input));
+
+      // Publish camera info
+      mCameraInfoMsg.header.stamp = stamp;
+      mPublisherCameraInfo->publish(mCameraInfoMsg);
+    } else {
+      deinitV4l2();
+      return;
     }
 
+    // Publish RAW RGB image if needed
     if (mPublisherRaw->get_subscription_count() > 0) {
-      cv::Mat inputMat(mCapturedImage.height, mCapturedImage.width, CV_8UC2, mCapturedImage.data);
-      cv::Mat outputMat;
-      // COLOR_YUV2BGR_Y422
-      cv::cvtColor(inputMat, outputMat, cv::COLOR_YUV2BGR_YUY2);
+      initRgb();
 
+      pipuck_mmal_convert(&mPipuckMmalRgb);
       auto message = sensor_msgs::msg::Image();
       message.encoding = "rgb8";
-      message.width = mCapturedImage.width;
-      message.height = mCapturedImage.height;
-      message.step = mCapturedImage.width * 3;
+      message.width = mPipuckMmalRgb.output.width;
+      message.height = mPipuckMmalRgb.output.height;
+      message.step = mPipuckMmalRgb.output.width * 3;
       message.is_bigendian = false;
-      message.header.stamp = now();
+      message.header.stamp = stamp;
       message.header.frame_id = "pipuck_image_raw";
-      message.data.assign(outputMat.data, outputMat.data + mCapturedImage.height * mCapturedImage.width * 3);
-    }
+      message.data.assign(mPipuckMmalRgb.output.data, mPipuckMmalRgb.output.data + mPipuckMmalRgb.output.size);
+      mPublisherRaw->publish(message);
+    } else
+      deinitRgb();
 
+    // Publish JPEG compressed image if needed
     if (mPublisherCompressed->get_subscription_count() > 0) {
-      if (!mJpegInitialized) {
-        pipuck_jpeg_init(&mCapturedImage, &mCompressedImage);
-        mJpegInitialized = true;
-      }
-      pipuck_jpeg_encode(&mCapturedImage, &mCompressedImage);
+      initJpeg();
 
+      // It's the same data for both topics
+      mPipuckMmalJpeg.input.data = mPipuckMmalRgb.input.data;
+      mPipuckMmalJpeg.input.size = mPipuckMmalRgb.input.size;
+
+      pipuck_mmal_convert(&mPipuckMmalJpeg);
       auto message = sensor_msgs::msg::CompressedImage();
       message.format = "jpeg";
-      message.header.stamp = now();
+      message.header.stamp = stamp;
       message.header.frame_id = "pipuck_image_compressed";
-      message.data.assign(mCompressedImage.data, mCompressedImage.data + mCompressedImage.size);
-
+      message.data.assign(mPipuckMmalJpeg.output.data, mPipuckMmalJpeg.output.data + mPipuckMmalJpeg.output.size);
       mPublisherCompressed->publish(message);
-    } else if (mJpegInitialized) {
-      pipuck_jpeg_deinit();
+    } else
+      deinitJpeg();
+  }
+
+  void initV4l2() {
```

Comment:
```suggestion
  void initV4l2IfNeeded() {
```

This way it shows that it is not a problem calling this function if it is already initialized.